### ISSUE TITLE
Support configurations where the machine can't boot from the target device

### ISSUE THREAD

**Issue by htejun** [iocost-benchmark/resctl-demo-image-recipe#18]
It turns out some machines have BIOS restrictions which don't allow booting from some storage devices. Installation works just fine but the boot entry doesn't show up, which is, according to the hw folks, because BIOS doesn't support booting from certain device at all.

Given that the machines can boot from USB and the device is fully accessible, I wonder whether having a boot option in the USB image to pivot to the installed system would work. e.g. Boot from the flash, show the users the devices with the image installed and pivot and continue booting there.

@obbardc what do you think? Any better ideas?

**Comment by htejun** [comment]
So, one way to implement this without changing things too much would be providing an option to pivot into the target root filesystem at the end of installation. Provided that the usb installation media uses the same kernel as the installed system, it can just offer to change root fs to the just installed filesystem and then fake-boot into it.

**Comment by santoshmahto7** [comment]
@obbardc @htejun 
Unfortunately flasher kernel(debian: linux-image-amd64) is different than installed kernel (https://github.com/iocost-benchmark/resctl-demo-linux). So pivoting the rootfs to installed system won't be useful.

But we can do pivoting/switching to installed system at bootloader level. That mean as:

- Insert flasher disk and boot from that.
- After bootup it will show UI to choose disks on which image will be installed. Here New option will be added which will be something like: "ChainLoad to boot from installed device in next boot".
- Once "Chainload to Boot from ...." is selected, it will setup flasher bootup configuration to chainload installed image bootloader  OR  flasher bootloader may be setup to load installed kernel and initramfs.
- After this is done and again system is booted from flasher device, flasher bootloader will load installed bootloader and then continue to load installed kernel and mount rootfs.


Another solution might be to use same kernel in both flasher and installed image but I am not sure of risk associated with it. In this case, pivoting rootfs will be easier since we already have rootfs pivot  using the kernel cmd params
`resctldemo.forcepivot`. we may need to adapt it for our situation.

**Comment by obbardc** [comment]
@santoshmahto7 sorry it has taken so long to reply to this.

I think this approach may be simpler:
- Install the custom kernel into the installer image instead of the Debian kernel (this also has the benefit of having the same device support in the installer) & test for regressions in qemu
- Add another menu option to the installer called "Pivot to rootfs" which would just list all found device paths & the device name (e.g. `/dev/sda1 Samsung 980 SSD`) in the first instance. We could do something like check the device label is the same as what the installer sets it to, even set a new label in the root image like "iocost-rootfs" and only show those for some operator sanity.

What do you think about that suggestion @santoshmahto7 ?

**Comment by santoshmahto7** [comment]
> @santoshmahto7 sorry it has taken so long to reply to this.
> 
> I think this approach may be simpler:
> 
> * Install the custom kernel into the installer image instead of the Debian kernel (this also has the benefit of having the same device support in the installer) & test for regressions in qemu
> * Add another menu option to the installer called "Pivot to rootfs" which would just list all found device paths & the device name (e.g. `/dev/sda1 Samsung 980 SSD`) in the first instance. We could do something like check the device label is the same as what the installer sets it to, even set a new label in the root image like "iocost-rootfs" and only show those for some operator sanity.
> 
> What do you think about that suggestion @santoshmahto7 ?

yes, if we are ready to replace debian kernel with resctl-demo kernel then there shouldn't be any issue.

Although another simpler option (without replacing debian kernel) is  to show boot menu that will load kernel( or initramfs) from target(installed) device.

Idea is that, first time the only bootmenu choices shown will be "Install resctl-demo"  (as it does currently). Once installation is done on target device and booted again from flasher device, another option will be shown something like " Boot from installed image" which will load installed kernel/initramfs. 

I am evaluating both option right now. Will let you know once done.

**Comment by obbardc** [comment]
I think replacing kernel with iocost kernel would be the simplest approach. Please do reach out via chat if you need support :-).

**Comment by santoshmahto7** [comment]
Unfortunately using kexec to switch kernel didn't work. Machine is hanged when kexec is called.
I could get is passed it once after following https://wiki.archlinux.org/title/kexec#Troubleshooting but still overall impression is that it's not reliable. Even in desktop linux, switching same kernel with kexec result in hanged state.
So need to look for other option

**Comment by ewyler** [comment]
kexec might still work in fact, so continuing that investigation

**Comment by santoshmahto7** [comment]
PR created here: https://github.com/iocost-benchmark/resctl-demo-image-recipe/pull/24
It would be nice if it could be tested with actual unbootable nvram

**Comment by ewyler** [comment]
Mostly dependent on us testing this and confirming it's good

**Comment by ewyler** [comment]
Santosh is working on patch to include ext4 into resctl build to make this work

**Comment by santoshmahto7** [comment]
This original issue is fixed with the mentioned PR, i.e flasher image boot properly now. 

But it seems, image installation on target device still have some issue and also kexec fails. The reason is that it fails to load some other required module specifically btrfs, vfat, nls_cp437 and thus fails to mount the target device when required to set some settings.

I have created another pull request to address this issue.
https://github.com/iocost-benchmark/resctl-demo-linux/pull/5
@htejun @obbardc

**Comment by santoshmahto7** [comment]
The relevant  fix is merged in resctl-demo-linux branch. Now pivoting to installed image works with latest flasher image as tested by me.
So this issue is also fixed.
